packages/firestore/src/remote/persistent_stream.ts

```diff
@@ -386,98 +395,86 @@ export abstract class PersistentStream<
+          // Normally we'd have to schedule the callback on the AsyncQueue.
+          // However, the following calls are safe to be called outside the
+          // AsyncQueue since they don't chain asynchronous calls
+          this.startStream(token);
+        }
       },
       (error: Error) => {
-        this.queue.enqueue(async () => {
-          if (this.state !== PersistentStreamState.Stopped) {
-            // Stream can be stopped while waiting for authorization.
-            const rpcError = new FirestoreError(
-              Code.UNKNOWN,
-              'Fetching auth token failed: ' + error.message
-            );
-            return this.handleStreamClose(rpcError);
-          }
+        dispatchIfStateUnchanged(() => {
+          const rpcError = new FirestoreError(
+            Code.UNKNOWN,
+            'Fetching auth token failed: ' + error.message
+          );
+          return this.handleStreamClose(rpcError);
         });
       }
     );
   }
 
   private startStream(token: Token | null): void {
-    if (this.state === PersistentStreamState.Stopped) {
-      // Stream can be stopped while waiting for authorization.
-      return;
-    }
-
     assert(
-      this.state === PersistentStreamState.Auth,
-      'Trying to start stream in a non-auth state'
+      this.state === PersistentStreamState.Starting,
+      'Trying to start stream in a non-starting state'
     );
-    // Helper function to dispatch to AsyncQueue and make sure that any
-    // close will seem instantaneous and events are prevented from being
-    // raised after the close call
-    const dispatchIfStillActive = (
-      stream: Stream<SendType, ReceiveType>,
-      fn: () => Promise<void>
-    ) => {
-      this.queue.enqueue(async () => {
-        // Only raise events if the stream instance has not changed
-        if (this.stream === stream) {
-          return fn();
-        }
-      });
-    };
 
-    // Only start stream if listener has not changed
-    if (this.listener !== null) {
-      const currentStream = this.startRpc(token);
-      this.stream = currentStream;
-      this.stream.onOpen(() => {
-        dispatchIfStillActive(currentStream, () => {
-          assert(
-            this.state === PersistentStreamState.Auth,
-            'Expected stream to be in state auth, but was ' + this.state
-          );
-          this.state = PersistentStreamState.Open;
-          return this.listener!.onOpen();
-        });
+    let dispatchIfStateUnchanged = this.stateGuardedDispatcher();
+
+    this.stream = this.startRpc(token);
+    this.stream.onOpen(() => {
+      dispatchIfStateUnchanged(() => {
+        assert(
+          this.state === PersistentStreamState.Starting,
+          'Expected stream to be in state Starting, but was ' + this.state
+        );
+        this.updateState(PersistentStreamState.Open);
+        // Need to recreate our dispatcher since we changed the state.
```

Comment:
This is the wart I mentioned...

Reply:
Ahh, yeah. That is better. I implemented this but called it closeCount (used with the now renamed `closeGuardedDispatcher`) because I think that's more precise (we increment it in close()) and understandable (what's a "stream generation"?)...  but if you don't like the naming, let me know.